### ISSUE TITLE
problem with telemetry data

### ISSUE THREAD

**Issue by basimolimajeed** [Azure/azure-iot-remote-monitoring#123]
Hi,
Occasionally the telemetry data history does not show (it seems that the display gets stuck on the original axes settings). This happened to me a number of times after restarting the web app. Any ideas why?

Thanks
![dashboardproblem](https://cloud.githubusercontent.com/assets/16608331/13459722/8a3a08ee-e06d-11e5-9043-df2e3325c8c6.png)

**Comment by networkfusion** [comment]
I also get intermittent display of telemetry. Sometimes it is stable but a lot of the time it displays for a short period and then shows the same image as above for another period. This only seems to occur with real devices and not the sample so I am wondering if it is getting confused by parsing a value of the wrong case...

**Comment by aguilaaj** [comment]
Can both of you provide more information on the devices you're using (and the SDK if you're using [Azure IoT SDKs](https://github.com/Azure/azure-iot-sdks)? If you don't want to go into depth here, feel free to send us a mail at iotsolhelp@microsoft.com

**Comment by basimolimajeed** [comment]
In my case I do not think it is to do with the sensor because both the alarms and the dial displays are showing the correct data, it is just the history chart that looks empty. But for information I am using a Texas Instruments Sensor Tag connected to PC on bluetooth and the PC is running the communication with Azure. One thing I did that seemed to have made the data appear is to restart the telemetry Stream Analytics job, but I honestly do not know if this was just a coincidence or it is something related. In any case the data in the chart disappeared again after an hour or so.

**Comment by networkfusion** [comment]
Currently my version of Remote monitoring is 603173f088f910cc96d57c6c38bd498f6ef7dce7 (master branch, 6th Feb). I am using Azure SDK 2.7.1.

The devices are NetMF. I am using a custom connection based on AmqpNetLite as the Azure IoT SDK does not support AMQP with NetMF (even though it is the most highly voted feature request). however I have copied the command  and telemetry structures from the remote monitoring examples within the SDK.

If I monitor the eventhub using Device explorer, I can see the data getting there as expected. I will however check the case of all parameters with the sample device once my monthly MSDN credit reactivates my Azure account #93.

**Comment by networkfusion** [comment]
I have checked the case of all parameters and they are all correct, so I am not sure where the problem lies. I am willing to let a member of the team delve into my instance if they wish.

**Comment by jamesweb-ms** [comment]
Try out change #126, the issue you running is Stream Analytics caches the case from the first message it sees.  Talking internally, the consensus is case shouldn't matter as you will run into issue with two differ

**Comment by jamesweb-ms** [comment]
Oops, hit wrong button...

**Comment by jamesweb-ms** [comment]
As I was sating, you run into issues with two different properties with same name different case... We shouldn't allow that.

**Comment by networkfusion** [comment]
@jamesweb-ms I have tried applying #126 and I am still seeing the same issue. If I get a chance I will do a screen capture.

**Comment by networkfusion** [comment]
As promised, here is a video of the problem occurring: https://www.youtube.com/watch?v=_CYteNalPIY

**Comment by princjef** [comment]
@networkfusion #126 was merged into master this evening so if you were checking against master before can you try it again?

Can you check out these two things and let us know what you find?

 1. Open up your F12 tools and make sure that your application is sending a request for `/api/v1/telemetry/dashboardDevicePane` every few seconds (default configuration is every three seconds) and that when you don't see any data on the graph that you are getting back an empty array for the "deviceTelemetryModels" property in the response, like the screenshot below. I would be surprised if this wasn't the case, but just want to make sure.
![empty array](https://cloud.githubusercontent.com/assets/3675357/13865287/e68a627e-ec66-11e5-80a3-72153505df33.png)

 2. Navigate to your subscription in the Azure Portal, select the resource group for your remote monitoring instance and navigate into the blobs for your Table Storage resource (specifically deviceTelemetry > deviceTelemetry). Drill through the dates to the folder for the hour where you ran into the issue (note that times are UTC). If you see more than one file in the folder and they have different schemas (you can download each file to check), that may be the issue.
![portal navigate](https://cloud.githubusercontent.com/assets/3675357/13865650/bf5abcea-ec6a-11e5-800c-1aeaa808c5b2.png)
![multiple blobs](https://cloud.githubusercontent.com/assets/3675357/13865667/f651a9b6-ec6a-11e5-8cd8-2878e49cc917.png)

**Comment by networkfusion** [comment]
I am indeed seeing empty deviceTelemetryModels":[] when data is not displayed in the dashboard.
The blob storage does contain multiple files (one for each device ID) with the only difference being the order of the headers, i.e.
                                                  
DeviceId
Temperature
Humidity
ExternalTemperature
EventProcessedUtcTime
PartitionId
EventEnqueuedUtcTime
IoTHub

VS


Temperature
ExternalTemperature
DeviceId
Humidity
EventProcessedUtcTime
PartitionId
EventEnqueuedUtcTime
IoTHub

Would that make a difference? If so, I can reorder the json on the device to match, but surely this shouldn't cause an issue...

**Comment by jamesweb-ms** [comment]
Yes, unfortunately it looks like Stream Analytics will generate new blobs if you change the column order (or add/remove columns).  The code as it currently exists, only reads the latest blob.  For now, can you align your column data so its consistent (you should see one blob with data from all devices)?  We are looking at options to handle both cases going forward.

**Comment by networkfusion** [comment]
After looking at my device code it is near impossible to order the properties in the same way as the simulator (Reliance on Hashtable serialized to Json in NetMF). I guess that leaves me with modifying the Telemetry SA query or stopping the simulated devices. I sure hope that you can come up with an official solution soon (but probably too late for my presentation to managers on Monday)...

**Comment by networkfusion** [comment]
Ok. I have modified the telemetry job to be the following:

`WITH 
    [StreamData]
AS (
    SELECT
        *
    FROM 
      [IoTHubStream] 
    WHERE
        [ObjectType] IS NULL -- Filter out device info and command responses
) 

SELECT
 -- *
   DeviceId, 
   Temperature, 
   Humidity, 
   ExternalTemperature, 
   EventProcessedUtcTime, 
   PartitionId, 
   EventEnqueuedUtcTime, 
   IoTHub -- Temporary work arround to make sure order is always the same for saving in blob storage
INTO
    [Telemetry]
FROM
    [StreamData]

SELECT
    DeviceId,
    AVG (Humidity) AS [AverageHumidity], 
    MIN(Humidity) AS [MinimumHumidity], 
    MAX(Humidity) AS [MaxHumidity], 
    5.0 AS TimeframeMinutes 
INTO
    [TelemetrySummary]
FROM
    [StreamData]
WHERE
    [Humidity] IS NOT NULL
GROUP BY
    DeviceId, 
    SlidingWindow (mi, 5)`

This cures the issue as a workaround! Roll on dynamic telemetry!

**Comment by aguilaaj** [comment]
@basimolimajeed - can you do the same and validate if your issue is resolved?

@networkfusion can you send us a mail at iotsolhelp@microsoft.com? I'd love to hear more about your use case for expanding RM and we really appreciate the good feedback you've been providing so far.

**Comment by networkfusion** [comment]
@aguilaaj I will do (but probably not until next week as now).

**Comment by princjef** [comment]
@networkfusion I have a code fix that addresses this issue. If you try to pull from the `jepri_multi-blob` branch and re-run, you should be able to pass in your columns in any order without issue.

**Comment by networkfusion** [comment]
@princjef I will after Monday, lets not upset the demo gods ;-)

**Comment by networkfusion** [comment]
@princjef Also I am liking the dynamic telemetry branch... is that in a usable state?

Also all, feel free to burn some of the many branches that have already been integrated, it is getting a bit messy and hard to see what is currently being worked on...

**Comment by jamesweb-ms** [comment]
@networkfusion Ariel just did a cleanup (we were actually talking about this morning).  The dynamic telemetry branch is experimental at this stage (Jeff is still working on some issues).  This is something we would like to support, but as you have noticed, there are some constraints that we are running into with how Stream Analytics generates the output files.  That is how Jeff knew about the multiple blobs :)

**Comment by jamesweb-ms** [comment]
@networkfusion @basimolimajeed if either of you can confirm that @princjef fix resolves your original issue (see message above for branch name), we will go ahead and merge and close this pull request.  You can do that next week once demo gods are appeased.

**Comment by networkfusion** [comment]
@aguilaaj I tried sending you an email to that address but it sent me one back saying I am not whitelisted?! is there another one?

**Comment by basimolimajeed** [comment]
Hi, unable to do much as the demo is in demand at the moment. I will do as soon as practically possible. sorry about the delay.

**Comment by networkfusion** [comment]
@aguilaaj did you read my last post about the email
@jamesweb-ms I should be able to update my instance tomorrow to test this and will let you know once complete.

**Comment by basimolimajeed** [comment]
Hi, Finally managed to deploy the jepri_multi-blob branch few hours ago. Seems to be working fine. Can we leave the issue open for a couple of days to see if it continues without problems?

**Comment by aguilaaj** [comment]
Fix merged to master, see #130.
@basimolimajeed - closing for now, feel free to reopen if need be.

**Comment by aguilaaj** [comment]
@networkfusion - got your mail, thanks! Do let us know once you've finished testing today.

**Comment by networkfusion** [comment]
I have now deployed the latest master (after problems with expired cached credentials in the deployment script) and multiple blobs seem to be showing up correctly on the dashboard. It is ashame that ASA doesn't support an "order alphabetically" statement as well though.
